### PR TITLE
Textcodec fixes

### LINKED ISSUE #1685: Improve performance of TextCodec#isDefinedAt
`TextCodec#isDefinedAt` needs to be optimized in a performant fashion for all subtypes. Currently, the method uses a slow implementation for the subtypes `UUIDCodec` and `BooleanCodec`. 

These implementations should be rewritten so that they do not allocate any memory (directly or indirectly), but rather validate that a string is a UUID or boolean by step-by-step iteration through characters in the string.

Once `isDefinedAt` is optimized for these subtypes, then the `decode` method of `TextCodec` can be rewritten to use a combination of `isDefinedAt` and `apply`.

### PR DESCRIPTION
**Pull request by Wosin**
Closes #1685.

Open to propositions on how to make UUID verification prettier.

### CODE REVIEW

File: zio-http/src/test/scala/zio/http/api/internal/UUIDCodecSpec.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      test("Should fail to validate String with leading hyphen") {
+        assertTrue(!UUIDCodec.isDefinedAt("-" + correctUUID))
+      },
+      test("Should work if UUID has only 0s") {
+        assertTrue(UUIDCodec.isDefinedAt("00000000-0000-0000-0000-000000000000"))
+      },
+      test("Should fail if UUID has incorrect character") {
+        assertTrue(!UUIDCodec.isDefinedAt("0000000Z-0000-0000-0000-000000000000"))
+      },
+      test("Should fail if UUID has two hyphens") {
```
Comment (bburdiliak) [inline zio-http/src/test/scala/zio/http/api/internal/UUIDCodecSpec.scala:31]: I would replace the illegal character Z with a legal one to really test the double hyphen case.

**Review by bburdiliak — CHANGES_REQUESTED** [review]

File: zio-http/src/test/scala/zio/http/api/internal/UUIDCodecSpec.scala (hunk 1 above)
Comment (Wosin) [reply zio-http/src/test/scala/zio/http/api/internal/UUIDCodecSpec.scala:31]: Oh, great catch. Will do!

File: zio-http/src/main/scala/zio/http/api/internal/TextCodec.scala [hunk 2]
```diff
@@ -104,7 +104,7 @@ private[api] object TextCodec {
     def encode(value: Boolean): String = value.toString
 
     // TODO: Make faster by hand-writing validation:
-    def isDefinedAt(value: String): Boolean = decode(value).isDefined
+    def isDefinedAt(value: String): Boolean = value == "1" || value == "0" || value == "true" || value == "false"
```
Comment (jdegoes) [inline zio-http/src/main/scala/zio/http/api/internal/TextCodec.scala:107]: This does not line up with `apply` since `apply` also accepts yes/on, no/off.

**Comment by jdegoes** [comment]
@Wosin Would love to get this in. Please just re-open when you have a chance to make the requested changes!

**Pull request closed without merging**